### PR TITLE
fix: correct mapping of DCAT themes to groups

### PR DESCRIPTION
**Pull request by jguo144**
# Description
Fixes Data.gov harvest reliability around groups and datastore upload, and tightens group handling in config processors.

## Changes

- Data.gov themes → groups: DCAT themes are no longer written onto package_dict['groups'] in datagov_to_ckan. Instead, DataGovHarvester builds source_dict['groups'] via _source_groups_from_datagov_themes() so RemoteGroups can resolve memberships using slug + human-readable title.
- RemoteGroups: Iterates source_dict.get('groups', []) so missing groups does not raise KeyError and deduplicate RemoteGroups by slug.
- Data.gov xloader: Implements a local upload_resources_to_datastore(context, package_dict) that only submits resources to xloader; it does not call push_data_dictionary (not applicable to Data.gov Search API payloads).
- Package search harvester: Uses source_dict.get('resources', []) / package_dict.get('resources', []) when iterating so None cannot be iterated.
- Tests: Adds DefaultGroups cases for merging defaults with existing groups (name-only) and skipping duplicates.

### CODE REVIEW

File: ckanext/custom_harvest/harvesters/datagov.py [hunk 1]
```diff
@@ -416,7 +449,7 @@ def import_stage(self, harvest_object):
                 if upload_to_datastore and p.plugin_loaded('xloader'):
                     # Get package dict again in case there's new resource ids
                     pkg_dict = p.toolkit.get_action('package_show')(package_context, {'id': package_id})
-                    upload_resources_to_datastore(package_context, pkg_dict, source_dict, base_search_url)
+                    upload_resources_to_datastore(package_context, pkg_dict)
```
Comment (nolanpruett) [inline ckanext/custom_harvest/harvesters/datagov.py]: Low priority: `base_search_url` is no longer utilized and is extracted on line 333. We should clean this up for clarity.

File: ckanext/custom_harvest/harvesters/datagov.py [hunk 2]
```diff
@@ -12,16 +12,51 @@
...
-    copy_across_resource_ids,
-    upload_resources_to_datastore
-)
+from ckanext.custom_harvest.harvesters.package_search import copy_across_resource_ids
 
 
 log = logging.getLogger(__name__)
 
 
+def upload_resources_to_datastore(context, package_dict):
```
Comment (nolanpruett) [inline ckanext/custom_harvest/harvesters/datagov.py]: Low priority suggestion: This has the same name as `package_search.upload_resources_to_datastore` but silently drops the data-dictionary push. Consider renaming to something like `_submit_resources_to_xloader` so it's clear these are different functions — right now a future reader (or a misplaced import) could easily confuse the two.

File: ckanext/custom_harvest/harvesters/datagov.py [hunk 3]
```diff
@@ -12,16 +12,51 @@
...
+                pass
+
+
+def _source_groups_from_datagov_themes(source_dict):
+    '''
+    Build source_dict-style group entries from DCAT themes for RemoteGroups.
+    Themes only (does not preserve any API ``groups`` on source_dict).
+    '''
+    dcat = source_dict.get('dcat') or {}
+    themes = set()
```
Comment (nolanpruett) [inline ckanext/custom_harvest/harvesters/datagov.py:44]: Also low priority fix: `themes` is a `set`, so `source_dict['groups']` comes out in arbitrary order each run. More importantly, if two themes munge to the same slug (e.g. "Economy" and "economy" both → "economy"), this produces duplicate entries in the returned list. `RemoteGroups` doesn't deduplicate those before extending `package_dict['groups']`. Consider building a `dict` keyed on `group_name` instead of a `list`, then returning `list(groups.values())`.

File: ckanext/custom_harvest/harvesters/datagov.py [hunk 4]
```diff
@@ -12,16 +12,51 @@
...
+                    'resource_id': resource.get('id'),
+                    'ignore_hash': False
+                }
+                p.toolkit.get_action('xloader_submit')(context, xloader_dict)
+            except p.toolkit.ValidationError as e:
+                log.debug(e)
+                pass
+
+
+def _source_groups_from_datagov_themes(source_dict):
```
Comment (nolanpruett) [inline ckanext/custom_harvest/harvesters/datagov.py:38]: Nit: Could we add tests for this function?

There are currently no tests for this function. It's the core of the architectural change, so worth covering the edge cases: themes present in both `source_dict['theme']` and `dcat['theme']` (set dedup), a `None` theme, a theme that munges to an empty string, and two themes that munge to the same slug, etc.

File: ckanext/custom_harvest/harvesters/datagov.py [hunk 5]
```diff
@@ -12,16 +12,51 @@
...
-)
+from ckanext.custom_harvest.harvesters.package_search import copy_across_resource_ids
 
 
 log = logging.getLogger(__name__)
 
 
+def upload_resources_to_datastore(context, package_dict):
+    '''Submit tabular resources to xloader (no remote datastore field hints).'''
+    for resource in package_dict.get('resources') or []:
```
Comment (nolanpruett) [inline ckanext/custom_harvest/harvesters/datagov.py]: Small stylistic nit: `package_dict.get('resources') or []` will also trigger when the key exists but holds an empty list. Use `package_dict.get('resources', [])` to match the style everywhere else (including the fix on `package_search.py:458`).

File: ckanext/custom_harvest/harvesters/datagov.py [hunk 6]
```diff
@@ -12,16 +12,51 @@
...
+            try:
+                log.info('Submitting harvested resource {0} to be xloadered'.format(resource.get('id')))
+                xloader_dict = {
+                    'resource_id': resource.get('id'),
+                    'ignore_hash': False
+                }
+                p.toolkit.get_action('xloader_submit')(context, xloader_dict)
+            except p.toolkit.ValidationError as e:
+                log.debug(e)
+                pass
```
Comment (nolanpruett) [inline ckanext/custom_harvest/harvesters/datagov.py]: nit: Maybe we have this here for linting purposes or something like that, but the `pass` at the end of the `except` block doesn't do anything

File: ckanext/custom_harvest/harvesters/datagov.py (hunk 1 above)
Comment (jguo144) [reply ckanext/custom_harvest/harvesters/datagov.py]: I removed the unnecessary `base_search_url`.

File: ckanext/custom_harvest/harvesters/datagov.py (hunk 2 above)
Comment (jguo144) [reply ckanext/custom_harvest/harvesters/datagov.py]: I've renamed the function to `_submit_resources_to_xloader`

File: ckanext/custom_harvest/harvesters/datagov.py (hunk 3 above)
Comment (jguo144) [reply ckanext/custom_harvest/harvesters/datagov.py:44]: I've updated RemoteGroups to deduplicate based on the slug

File: ckanext/custom_harvest/harvesters/datagov.py (hunk 4 above)
Comment (jguo144) [reply ckanext/custom_harvest/harvesters/datagov.py:38]: Added tests for _source_groups_from_datagov_themes

File: ckanext/custom_harvest/harvesters/datagov.py (hunk 5 above)
Comment (jguo144) [reply ckanext/custom_harvest/harvesters/datagov.py]: I've updated it to use `package_dict.get('resources', [])` and `dcat = source_dict.get('dcat', {})`

File: ckanext/custom_harvest/harvesters/datagov.py (hunk 6 above)
Comment (jguo144) [reply ckanext/custom_harvest/harvesters/datagov.py]: I've removed the unnecessary use of `pass`

**Review by nolanpruett — APPROVED** [review]
